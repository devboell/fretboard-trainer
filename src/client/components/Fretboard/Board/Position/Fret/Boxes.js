import { css } from 'styled-components'
import SkinWrapper from './SkinWrapper'

const Boxes = SkinWrapper.extend`
  height: 80%;
  width: 90%;
  margin: 10% 5%;
  border: 1px solid darkgray;
  border-radius: 3px;
  ${props => props.isHighlighted && css`background-color: ${props.theme.highlight};`}
`

export default Boxes
