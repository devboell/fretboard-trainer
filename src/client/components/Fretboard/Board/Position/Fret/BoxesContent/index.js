import React from 'react'
import pt from 'prop-types'
import Wrapper from './Wrapper'


const BoxesContent = ({ note }) =>
  <Wrapper>
    {note}
  </Wrapper>

BoxesContent.propTypes = {
  note: pt.string.isRequired,
}

export default BoxesContent
