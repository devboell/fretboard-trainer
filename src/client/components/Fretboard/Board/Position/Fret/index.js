import React from 'react'
import pt from 'prop-types'
import BoxesFret from './BoxesFret'
import StringsFret from './StringsFret'


import Wrapper from './Wrapper'

/* eslint-disable react/prefer-stateless-function */
class Fret extends React.Component {
  render() {
    const { type, showNotes, isHighlighted } = this.context
    const Outer = type === 'boxes' ? BoxesFret : StringsFret

    return (
      <Wrapper>
        <Outer
          note={this.props.note}
          showNotes={showNotes}
          isHighlighted={isHighlighted}
        />
      </Wrapper>)
  }
}


Fret.propTypes = {
  note: pt.string.isRequired,
}

Fret.contextTypes = {
  type: pt.string.isRequired,
  showNotes: pt.bool.isRequired,
  isHighlighted: pt.bool.isRequired,
}

export default Fret
