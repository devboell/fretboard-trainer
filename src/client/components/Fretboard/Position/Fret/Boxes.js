import { css } from 'styled-components'
import SkinWrapper from './SkinWrapper'


const background = ({ theme, isHighlighted, status }) =>
  isHighlighted && css`background-color: ${theme.highlight[status]};`

const Boxes = SkinWrapper.extend`
  height: 80%;
  width: 90%;
  margin: 10% 5%;
  border: 1px solid darkgray;
  border-radius: 3px;
  ${props => background(props)}
`

export default Boxes
