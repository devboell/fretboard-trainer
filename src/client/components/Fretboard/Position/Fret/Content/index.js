import React from 'react'
import pt from 'prop-types'
// import { isNil } from 'ramda'
// import { Note } from 'tonal'
import Wrapper from './Wrapper'

/*
const formatSharp = sh => sh.replace('#', '\u266F')
const formatFlat = fl => fl.replace('b', '\u266D')

const formatEnharmonics = ([sh, fl]) =>
  `${formatSharp(sh)}/${formatFlat(fl)}`

const formatNote = (note, showEnharmonic) => {
  const enharmonic = Note.enharmonic(note)
  const hasEnharmonic = note !== enharmonic

  return showEnharmonic && hasEnharmonic
    ? formatEnharmonics([note, enharmonic])
    : formatSharp(note)
}
*/

/* the content width is a big problem, some options to exlore:
- format this way 'D#/Eb5'
- make ur own glyph
- showPcs/pitches option
*/

const Content = ({ content }) =>
  /*
  const showEnharmonic = true
  const isNote = !isNil(Note.name(content))
  const note = isNote
    ? formatNote(content, showEnharmonic)
    : undefined

  {note || content}
  */

  (
    <Wrapper>
      {content}
    </Wrapper>
  )


Content.propTypes = {
  content: pt.string.isRequired,
}

export default Content
