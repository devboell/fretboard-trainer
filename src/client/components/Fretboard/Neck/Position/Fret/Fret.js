import React from 'react'
import pt from 'prop-types'
import { isEmpty, isNil, equals, compose } from 'ramda'
import { withTheme } from 'styled-components'
import { Note } from 'tonal'
import { isEqual } from 'lib/tonal-helpers'
import { locShape, noteSelectionShape, locSelectionShape } from 'lib/shapes'
import { fretWrapper } from 'components/Fretboard/Skins/skins'

import Content from './Content/Content'
import Wrapper from './Wrapper'


/* eslint-disable react/prefer-stateless-function */
export class Fret extends React.Component {
  render() {
    const { note, loc, theme } = this.props
    const {
      skinType,
      showNotes,
      showOctaves,
      showSelection,
      showEnharmonics,
      selectedNotes,
      selectedLocations,
      clickAction,
    } = this.context

    const findNote = (acc, curr) => (
      isEqual(note)(curr.note)
        ? curr
        : acc
    )

    const findLoc = (acc, curr) => (
      equals(loc)(curr.loc)
        ? curr
        : acc
    )

    const noteSelection = selectedNotes.reduce(findNote, undefined)
    const locSelection = selectedLocations.reduce(findLoc, undefined)

    const selection = noteSelection || locSelection
    const isSelected = !isNil(selection)

    const content = compose(
      result => (showSelection && isSelected ? selection.label : result),
      result => (showNotes ? note : result),
    )('')
    const hasContent = !isEmpty(content)

    const color = compose(
      result => (isSelected ? theme.statusMap[selection.status] : result),
      result => (showOctaves ? theme.octaveMap[Note.oct(note)] : result),
    )('none')
    const isHighlighted = color !== 'none'

    const SkinWrapper = fretWrapper(skinType)

    return (
      <Wrapper onClick={() => clickAction(note, loc)}>
        <SkinWrapper {...{ isHighlighted, color }}>
          {hasContent &&
            <Content {...{ content, showEnharmonics }} />
          }
        </SkinWrapper>
      </Wrapper>)
  }
}


Fret.propTypes = {
  note: pt.string.isRequired,
  loc: locShape.isRequired,
  theme: pt.shape({}).isRequired,
}

Fret.contextTypes = {
  skinType: pt.string.isRequired,
  showNotes: pt.bool.isRequired,
  showOctaves: pt.bool.isRequired,
  showSelection: pt.bool.isRequired,
  showEnharmonics: pt.bool.isRequired,
  selectedNotes: pt.arrayOf(noteSelectionShape).isRequired,
  selectedLocations: pt.arrayOf(locSelectionShape).isRequired,
  clickAction: pt.func.isRequired,
}

export default withTheme(Fret)
