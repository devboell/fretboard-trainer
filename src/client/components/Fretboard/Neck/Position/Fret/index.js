import React from 'react'
import pt from 'prop-types'
import { isNil } from 'ramda'
import { withTheme } from 'styled-components'

import { locShape, noteSelectionShape, locSelectionShape } from 'lib/shapes'
import {
  selectedNote,
  selectedLoc,
  decideContent,
  decideColor,
} from 'lib/fret'
import { fretWrapper } from 'components/Fretboard/Skins/skins'

import Content from './Content'
import Wrapper from './Wrapper'


/* eslint-disable react/prefer-stateless-function */
export class Fret extends React.Component {
  render() {
    const { note, loc, theme } = this.props
    const {
      skinType,
      showNotes,
      showOctaves,
      showSelection,
      showEnharmonics,
      selectedNotes,
      selectedLocations,
      clickAction,
    } = this.context

    const noteSelection = selectedNote(note, selectedNotes)
    const locSelection = selectedLoc(loc, selectedLocations)
    const selection = noteSelection || locSelection


    const content = decideContent(
      showSelection,
      selection,
      showNotes,
      note,
    )

    const color = decideColor(
      selection,
      theme,
      showOctaves,
      note,
    )

    const isHighlighted = !isNil(color)

    const SkinWrapper = fretWrapper(skinType)

    return (
      <Wrapper onClick={() => clickAction(note, loc)}>
        <SkinWrapper {...{ isHighlighted, color }}>
          {content &&
            <Content {...{ content, showEnharmonics }} />
          }
        </SkinWrapper>
      </Wrapper>)
  }
}


Fret.propTypes = {
  note: pt.string.isRequired,
  loc: locShape.isRequired,
  theme: pt.shape({}).isRequired,
}

Fret.contextTypes = {
  skinType: pt.string.isRequired,
  showNotes: pt.bool.isRequired,
  showOctaves: pt.bool.isRequired,
  showSelection: pt.bool.isRequired,
  showEnharmonics: pt.bool.isRequired,
  selectedNotes: pt.arrayOf(noteSelectionShape).isRequired,
  selectedLocations: pt.arrayOf(locSelectionShape).isRequired,
  clickAction: pt.func.isRequired,
}

export default withTheme(Fret)
