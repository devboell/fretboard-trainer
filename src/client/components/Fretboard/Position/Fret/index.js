import React from 'react'
import pt from 'prop-types'
import { isEmpty, isNil, equals, compose } from 'ramda'
import { isEqual, oct } from 'lib/tonal-helpers'
import { locShape, noteSelectionShape, locSelectionShape } from 'lib/shapes'
import { fretWrapper } from 'components/Fretboard/skins'

import Content from './Content'
import Wrapper from './Wrapper'

const octStatusMap = {
  2: 'oct2',
  3: 'oct3',
  4: 'oct4',
  5: 'oct5',
}

const octStatus = note => octStatusMap[oct(note)]

/* eslint-disable react/prefer-stateless-function */
class Fret extends React.Component {
  render() {
    const { note, loc } = this.props
    const {
      skinType,
      showNotes,
      showOctaves,
      showSelection,
      selectedNotes,
      selectedLocations,
      clickAction,
    } = this.context

    const findNote = (acc, curr) => (
      isEqual(note)(curr.note)
        ? curr
        : acc
    )

    const findLoc = (acc, curr) => (
      equals(loc)(curr.loc)
        ? curr
        : acc
    )

    const noteSelection = selectedNotes.reduce(findNote, undefined)
    const locSelection = selectedLocations.reduce(findLoc, undefined)

    const selection = noteSelection || locSelection
    const isSelected = !isNil(selection)

    const content = compose(
      result => (showSelection && isSelected ? selection.label : result),
      result => (showNotes ? note : result),
    )('')
    const hasContent = !isEmpty(content)

    const status = compose(
      result => (isSelected ? selection.status : result),
      result => (showOctaves ? octStatus(note) : result),
    )('none')
    const isHighlighted = status !== 'none'

    const SkinWrapper = fretWrapper(skinType)

    return (
      <Wrapper onClick={() => clickAction(note, loc)}>
        <SkinWrapper {...{ isHighlighted, status }}>
          {hasContent &&
            <Content content={content} />
          }
        </SkinWrapper>
      </Wrapper>)
  }
}


Fret.propTypes = {
  note: pt.string.isRequired,
  loc: locShape.isRequired,
}

Fret.contextTypes = {
  skinType: pt.string.isRequired,
  showNotes: pt.bool.isRequired,
  showOctaves: pt.bool.isRequired,
  showSelection: pt.bool.isRequired,
  selectedNotes: pt.arrayOf(noteSelectionShape).isRequired,
  selectedLocations: pt.arrayOf(locSelectionShape).isRequired,
  clickAction: pt.func.isRequired,
}

export default Fret
