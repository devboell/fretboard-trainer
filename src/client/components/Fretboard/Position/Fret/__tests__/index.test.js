import React from 'react'

import Fret from '../index'


describe('Fret component', () => {
  const clickAction = jest.fn()

  const props = {
    note: 'E2',
    loc: { str: 5, pos: 0 },
  }
  const defaultContext = {
    skinType: 'boxes',
    showNotes: false,
    showOctaves: false,
    showSelection: false,
    selectedNotes: [],
    selectedLocations: [],
    clickAction,
  }

  it('defaultContext, no selections', () => {
    const wrapper = shallow(<Fret {...props} />, { context: defaultContext })
    expect(wrapper).toMatchSnapshot()
  })

  describe('-- selectedNotes & labels --', () => {
    it('showSelection=false, showNotes=false, note selected', () => {
      const selectedNotes = [{ note: 'E2', status: 'selected', label: 'root' }]
      const context = { ...defaultContext, selectedNotes }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showSelection=true, showNotes=false, note selected', () => {
      const selectedNotes = [{ note: 'E2', status: 'selected', label: 'root' }]
      const context = {
        ...defaultContext,
        selectedNotes,
        showSelection: true,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showSelection=true, showNotes=true, note selected', () => {
      const selectedNotes = [{ note: 'E2', status: 'selected', label: 'root' }]
      const context = {
        ...defaultContext,
        selectedNotes,
        showSelection: true,
        showNotes: true,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showSelection=false, showNotes=true, no note selected', () => {
      const context = {
        ...defaultContext,
        showSelection: false,
        showNotes: true,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showSelection=true, showNotes=true, note selected', () => {
      const selectedNotes = [{ note: 'E2', status: 'selected', label: 'root' }]
      const context = {
        ...defaultContext,
        selectedNotes,
        showSelection: false,
        showNotes: true,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('-- selectedNotes & highlights --', () => {
    it('showOctaves=false, note selected', () => {
      const selectedNotes = [{ note: 'E2', status: 'foo', label: 'root' }]
      const context = { ...defaultContext, selectedNotes }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showOctaves=true, note selected', () => {
      const selectedNotes = [{ note: 'E2', status: 'foo', label: 'root' }]
      const context = {
        ...defaultContext,
        showOctaves: true,
        selectedNotes,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showOctaves=true, no note selected', () => {
      const context = {
        ...defaultContext,
        showOctaves: true,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('-- selectedLocations & labels --', () => {
    it('showSelection=false, loc selected', () => {
      const loc = { str: 5, pos: 0 }
      const selectedLocations = [{ loc, status: 'foo', label: 'root' }]
      const context = { ...defaultContext, selectedLocations }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showSelection=true, loc selected', () => {
      const loc = { str: 5, pos: 0 }
      const selectedLocations = [{ loc, status: 'foo', label: 'root' }]
      const context = {
        ...defaultContext,
        showSelection: true,
        selectedLocations,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('-- selectedLocations & highlights --', () => {
    it('showOctaves=false, loc selected', () => {
      const loc = { str: 5, pos: 0 }
      const selectedLocations = [{ loc, status: 'foo', label: 'root' }]
      const context = { ...defaultContext, selectedLocations }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('showOctaves=true, loc selected', () => {
      const loc = { str: 5, pos: 0 }
      const selectedLocations = [{ loc, status: 'foo', label: 'root' }]
      const context = {
        ...defaultContext,
        showOctaves: true,
        selectedLocations,
      }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('-- other selection --', () => {
    it('should not show selection if selectedNote doesn\'t match', () => {
      const selectedNotes = [{ note: 'A2', status: 'selected', label: 'root' }]
      const context = { ...defaultContext, selectedNotes }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })

    it('should not show selection if selectedLocation doesn\'t match', () => {
      const loc = { str: 5, pos: 5 }
      const selectedLocations = [{ loc, status: 'foo', label: 'root' }]
      const context = { ...defaultContext, selectedLocations }
      const wrapper = shallow(<Fret {...props} />, { context })
      expect(wrapper).toMatchSnapshot()
    })
  })

  it('clickAction gets called with correct args', () => {
    const wrapper = shallow(<Fret {...props} />, { context: defaultContext })
    wrapper.simulate('click')
    expect(clickAction).toBeCalledWith('E2', { str: 5, pos: 0 })
  })
})
