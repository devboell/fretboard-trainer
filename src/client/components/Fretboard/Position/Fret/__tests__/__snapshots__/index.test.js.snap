// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fret component -- other selection -- should not show selection if selectedLocation doesn't match 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={false}
    status="none"
  />
</styled.div>
`;

exports[`Fret component -- other selection -- should not show selection if selectedNote doesn't match 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={false}
    status="none"
  />
</styled.div>
`;

exports[`Fret component -- selectedLocations & highlights -- showOctaves=false, loc selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="foo"
  />
</styled.div>
`;

exports[`Fret component -- selectedLocations & highlights -- showOctaves=true, loc selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="foo"
  />
</styled.div>
`;

exports[`Fret component -- selectedLocations & labels -- showSelection=false, loc selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="foo"
  />
</styled.div>
`;

exports[`Fret component -- selectedLocations & labels -- showSelection=true, loc selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="foo"
  >
    <Content
      content="root"
    />
  </styled.div>
</styled.div>
`;

exports[`Fret component -- selectedNotes & highlights -- showOctaves=false, note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="foo"
  />
</styled.div>
`;

exports[`Fret component -- selectedNotes & highlights -- showOctaves=true, no note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="oct2"
  />
</styled.div>
`;

exports[`Fret component -- selectedNotes & highlights -- showOctaves=true, note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="foo"
  />
</styled.div>
`;

exports[`Fret component -- selectedNotes & labels -- showSelection=false, showNotes=false, note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="selected"
  />
</styled.div>
`;

exports[`Fret component -- selectedNotes & labels -- showSelection=false, showNotes=true, no note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={false}
    status="none"
  >
    <Content
      content="E2"
    />
  </styled.div>
</styled.div>
`;

exports[`Fret component -- selectedNotes & labels -- showSelection=true, showNotes=false, note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="selected"
  >
    <Content
      content="root"
    />
  </styled.div>
</styled.div>
`;

exports[`Fret component -- selectedNotes & labels -- showSelection=true, showNotes=true, note selected 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="selected"
  >
    <Content
      content="root"
    />
  </styled.div>
</styled.div>
`;

exports[`Fret component -- selectedNotes & labels -- showSelection=true, showNotes=true, note selected 2`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={true}
    status="selected"
  >
    <Content
      content="E2"
    />
  </styled.div>
</styled.div>
`;

exports[`Fret component defaultContext, no selections 1`] = `
<styled.div
  onClick={[Function]}
>
  <styled.div
    isHighlighted={false}
    status="none"
  />
</styled.div>
`;
